Users/rahur/release individual driver version (#2090)
ni/nimi-python: CONTRIBUTING.md

Reviewer: How will Git tagging work for independent driver releases? Currently the tag is just a version, like 1.4.9. Will you add the driver name to the tag (like nidcpower-1.4.9) or will the tag version still apply to all drivers?

When we converted https://github.com/ni/measurement-plugin-python to a monorepo, we kept the global version tagging, but we don't release that repo's packages independently.

Reviewer 2: >nidcpower-1.4.9

Yes, exactly like that.
Sadly, the package name will show up in the version on ReadTheDocs, but it can't be helped.

This method will also allow us to update the stable version of the documentation.

We'll also have to do more work to release multiple packages because of the individual tags, but that's just the nature of separate versions.

        ```
    1. Ensure no commits are made on ni/nimi-python/master until the release is complete
    1. Create and checkout a branch for release-related changes
    1. Perform Version Bump (If Needed)
        * If you need to upgrade the major or minor versions, include any of the following parameters:
           * --increment-major-version - To increment the major version of package. This will update the version to (N+1).X.X.dev0
           * --increment-minor-version - To increment the minor version of package. This will update the version to X.(N+1).X.dev0

>             For example: 
>                 ```bash
>                 python3 tools/build_release.py --increment-minor-version
>                 ```

Reviewer 2: ```suggestion
           * Example: `python3 tools/build_release.py --increment-minor-version`
```

Markdown treats that "For example" as part of the previous line when rendering.()

        ```
    1. Ensure no commits are made on ni/nimi-python/master until the release is complete
    1. Create and checkout a branch for release-related changes
    1. Perform Version (major or minor) Bump (If Needed)

>         * If you need to upgrade the major, minor, patch or dev versions, include any of the following parameters:

Reviewer 2: ```suggestion
        * If you need to upgrade the major, minor or patch version, include one of the following parameters:
```

Author: Resolved

Reviewer 2: It still says "any" instead of "one", but it's fine. The error is clear enough.

           * --increment-major-version - To increment the major version of package. This will update the version to (N+1).X.X.dev0
           * --increment-minor-version - To increment the minor version of package. This will update the version to X.(N+1).X.dev0
           * --increment-patch-version - To increment the patch version of package. This will update the version to X.X.(N+1).dev0
            For example: 
                ```bash
                python3 tools/build_release.py --increment-minor-version
                ```
        * If you need to update the version for any specific driver(s), include the `drivers` parameter. By default, all drivers will be considered.

>         For example: 
>             ```bash
>             python3 tools/build_release.py --drivers nidcpower --increment-minor-version
>             ```

Reviewer 2: ```suggestion
          For example: 
              ```bash
              python3 tools/build_release.py --drivers nidcpower --increment-minor-version
              ```
```

Author: Resovled

Reviewer 2: Spacing is still a bit off, but it's close enough and looks good when rendered.

        * Commit to branch
    1. Update [CHANGELOG.md](./CHANGELOG.md)
        * Delete empty (i.e. No changes) sub-sections under "Unreleased" section
        * Change the "Unreleased" header to the version of the release
        * Change [Unreleased] in TOC to the version of the release
        * Commit to branch
    1. Update release versions
        * `python3 tools/build_release.py --update-for-release`
            * For each module, this will drop the .devN from our versions in config_addon.py and update the LATEST_RELEASE versions to match.

>             * If you need to release any specific module(s), include the `drivers` parameter. 
>             For example: 
>                 ```bash
>                 python3 tools/build_release.py --drivers nidcpower --update-for-release
>                 ```

Reviewer 2: ```suggestion
            * If you need to release any specific module(s), include the `drivers` parameter. 
              For example: 
                  ```bash
                  python3 tools/build_release.py --drivers nidcpower --update-for-release
                  ```
```

Author: Resolved

Reviewer 2: Spacing is still a bit off, but it's close enough and looks good when rendered.

        ```
    1. Ensure no commits are made on ni/nimi-python/master until the release is complete
    1. Create and checkout a branch for release-related changes
    1. Perform Version Bump (If Needed)
        * If you need to upgrade the major, minor, patch or dev versions, include any of the following parameters:
           * --increment-major-version - To increment the major version of package. This will update the version to (N+1).X.X.dev0
           * --increment-minor-version - To increment the minor version of package. This will update the version to X.(N+1).X.dev0

>            * --increment-patch-version - To increment the patch version of package. This will update the version to X.X.(N+1).dev0

Author: Even i feel so.. And we also will not use build number increment as well. (We may remove --increment-build-number option if fine).  as part of post release activity, we will be upgrading the patch number and appending dev0 version.  Before the final release we will give major or minor upgrade option if needed. If they do --increment-patch-version later, we can error.

>            * --increment-patch-version - To increment the patch version of package. This will update the version to X.X.(N+1).dev0

Author: Removed --increment-build-number option.

>            * --increment-patch-version - To increment the patch version of package. This will update the version to X.X.(N+1).dev0

Reviewer 2: Nitpick: I still don't like mention of --increment-patch-version here. I think we should just document that directly under post-release steps, because it's the only place we'll use it.

>            * --increment-patch-version - To increment the patch version of package. This will update the version to X.X.(N+1).dev0

Author: I have updated in post release and prerelease steps

    1. Perform Version Bump (If Needed)
        * If you need to upgrade the major, minor, patch or dev versions, include any of the following parameters:
           * --increment-major-version - To increment the major version of package. This will update the version to (N+1).X.X.dev0
           * --increment-minor-version - To increment the minor version of package. This will update the version to X.(N+1).X.dev0
           * --increment-patch-version - To increment the patch version of package. This will update the version to X.X.(N+1).dev0
           * --increment-build-number - To increment the development number of package. This will update the version to X.X.X.dev(N+1)
            For example: 
                ```bash
                python3 tools/build_release.py --increment-minor-version
                ```
        * If you need to update the version for any specific driver(s), include the `drivers` parameter. By default, all drivers will be considered.
        For example: 
            ```bash
            python3 tools/build_release.py --drivers nidcpower --increment-minor-version
            ```
        * Commit to branch
    1. Update [CHANGELOG.md](./CHANGELOG.md)

>         * Delete empty (i.e. No changes) sub-sections under "Unreleased" section for each drivers.

Reviewer 2: We should only edit the subsections for APIs that are releasing.

Author: Agreed Jay

Author: I have made changes in other PR for Changelog

    1. Wait until the pull request has been approved
    1. Upload the releases to PyPI
        * `python3 tools/build_release.py --upload`
        * If you need to upload any specific module(s), include the `drivers` parameter. 

>         For example: 
>              ```bash
>             python3 tools/build_release.py --drivers nidcpower --upload
>             ```

Reviewer 2: ```suggestion
          For example: 
              ```bash
              python3 tools/build_release.py --drivers nidcpower --upload
              ```
```

Author: Resolved

Reviewer 2: Spacing is still a bit off, but it's close enough and looks good when rendered.

        * You will need to type in your PyPI credentials
    1. Merge the pull request to origin/master

>     1. Create a release on GitHub using the portion from the changelog for this release for the description
>         * Add the ZIP files under `generated/examples` for each module as a release artifact.
>         * This should trigger the [check_latest_release](.github/workflows/check_latest_release.yml) workflow. Check the [results](https://github.com/ni/nimi-python/actions/workflows/check_latest_release.yml) before continuing.

Reviewer 2: ```suggestion
    1. For each package released, create a release on GitHub using the module's portion from the changelog for this release for the description
        * The release tag should be named as follows: `MODULE_NAME-version`.
          * Example: `nidcpower-1.5.0`.
          * This tag format allows the individual `Read the Docs` projects to determine whether a release applies to them.
        * Add the ZIP files under `generated/examples` for each module (not just the releasing one) as a release artifact.
          * Internal test code will only look for the latest release tag and expect it to have examples attached for any module
        * This should trigger the [check_latest_release](.github/workflows/check_latest_release.yml) workflow. Check the [results](https://github.com/ni/nimi-python/actions/workflows/check_latest_release.yml) before continuing.
```

Author: Resolved

    1. Wait until the pull request has been approved
    1. Upload the releases to PyPI
        * `python3 tools/build_release.py --upload`
        * If you need to upload any specific module(s), include the `drivers` parameter. 
        For example: 
             ```bash
            python3 tools/build_release.py --drivers nidcpower --upload
            ```
        * You will need to type in your PyPI credentials
    1. Merge the pull request to origin/master
    1. Create a release on GitHub using the portion from the changelog for this release for the description
        * Add the ZIP files under `generated/examples` for each module as a release artifact.
        * This should trigger the [check_latest_release](.github/workflows/check_latest_release.yml) workflow. Check the [results](https://github.com/ni/nimi-python/actions/workflows/check_latest_release.yml) before continuing.
1. Post-Release Steps
    1. Create and checkout another branch for post-release changes

>     1. Update the module versions. 

Reviewer 2: It's possible we could also bump the build version, but I don't think we'll be doing that, because I don't think we'll release dev builds, in the future.

>     1. Update the module versions. 

Author: I have tried to be specific about the major and minor version . So that the build and patch number upgrading will not be given importance. patch upgrade is already done as part of post release activity.

>     1. Update the module versions. 

Reviewer 2: This **is** post-release steps, so it should be the patch version. You should be specific about major and minor version in the pre-release steps.

>     1. Update the module versions. 

Author: I have updated in post release and pre release and tried to be specific. 

Do we need to deprecate the --increment-build-number option?  Since we are not using it much.

Suggestions:
As part of post-release activity, the patch version bumping is done. This also appends dev0 as the end of version.
And during pre-release, minor or major version can be bumped if needed. 
And --update-for-release is used to remove dev0 from the version across all pages.

>     1. Update the module versions. 

Author: Implemented above comment.

>     1. Update the module versions(major or minor) if required. 

Reviewer 2: ```suggestion
    1. Update the module versions (major or minor) if required.
```

Nitpick.
